### DISCUSSION TITLE
writableCoupledValue with NodalUO

### DISCUSSION THREAD

**Discussion by maxnezdyur** [idaholab/moose#22875] [Q&A General] [answered]
I am trying to use writableCoupledValue with a NodalUO. Below is an example code. In the input file, I set an auxvariable to an initial condition of 100 and have a simple NodalUO that should change all the values to 1(if I am using this correctly). Instead, the exodus file just shows that the aux variable stays at 100. My actual code has me changing a few variables at the beginning of a time step so that why I want to use the NodalUO.
```
[Mesh]
  [gen]
    type = GeneratedMeshGenerator
    dim = 2
    elem_type = QUAD4
  []
[]

[Problem]
  kernel_coverage_check = false
  material_coverage_check = false
  solve = false
[]

[AuxVariables]
  [change]
    initial_condition = 100
  []
[]

[Executioner]
  type = Transient
  solve_type = 'NEWTON'
  dt = 1
  num_steps = 2
[]

[UserObjects]
  [change_var]
    type = NodeUOWriteableTest
    variable_to_change = change
    execute_on = TIMESTEP_END
    force_preaux = true
  []
[]
[Outputs]
  [exo]
    type = Exodus
    file_base = nodal_uo_test/file
  []
[]

```
```
#include "NodeUOWriteableTest.h"

registerMooseObject("insectApp", NodeUOWriteableTest);

InputParameters
NodeUOWriteableTest::validParams()
{
  InputParameters params = NodalUserObject::validParams();
  params.addRequiredCoupledVar("variable_to_change", "");

  params.addClassDescription("");
  return params;
}

NodeUOWriteableTest::NodeUOWriteableTest(const InputParameters & parameters)
  : NodalUserObject(parameters), _variable_to_change(writableCoupledValue("variable_to_change"))
{
}

void
NodeUOWriteableTest::execute()
{
  _variable_to_change[_qp] = 1.0;
}

```
```
#pragma once

#include "MooseTypes.h"
#include "NodalUserObject.h"

/**
 */
class NodeUOWriteableTest : public NodalUserObject
{
public:
  static InputParameters validParams();

  NodeUOWriteableTest(const InputParameters & parameters);

  virtual void initialize() override{};
  virtual void execute() override;
  virtual void finalize() override{};
  virtual void threadJoin(const UserObject & /* uo*/) override{};
  virtual void initialSetup() override{};

private:
  VariableValue & _variable_to_change;
};

```

**Comment by dschwen** [comment, marked as answer]
Hm, looks like that API function is only tested from within an AuxKernel, not wuthin a NodalUO (where it would make a ton of sense!)

**Reply by dschwen** [reply]
My guess would be that the AuxSystem solution vector is not properly closed in or after the NodalUO loop.

**Reply by maxnezdyur** [reply]
Good to know, so there might a simple stopgap of closing the AuxSystem solution vector in `finalize()`. I prefer not to have to move this into an auxkernel. I'll see if I find anything.

**Reply by dschwen** [reply]
You'll need to do something like this: https://github.com/idaholab/moose/blob/next/framework/src/loops/ComputeNodalAuxVarsThread.C#L105-L110

which will require compiling a list of aux variables that can be written to.

**Reply by maxnezdyur** [reply]
Thanks for finding that snippet! 
I may directly change the values in the solution vector. For my use case, I need to change the "old" values of variables.  In `ElementSubdomainModifier::setOldAndOlderSolutionsForMovedNodes` it shows how an ElementUO can change the old values by directly modifying the entire solution vector. I created a member in `Coupleable` called `writableCoupledValueOld`, hoping it would work out of the box for variables, not just auxvariables. If you can think of any other way to change "old" values of variables let me know, but for the purpose of this discussion post, I will mark it answered.
